File: app/demo/src/main/java/net/gotev/recycleradapterdemo/MainActivity.java
package net.gotev.recycleradapterdemo;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import net.gotev.recycleradapter.RecyclerAdapter;

import java.util.Random;
import java.util.UUID;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class MainActivity extends AppCompatActivity {

    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;

    private RecyclerAdapter adapter;
    private Random random;

    private Random getRandom() {
        if (random == null) {
            random = new Random(System.currentTimeMillis());
        }
        return random;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        adapter = new RecyclerAdapter();
        adapter.setEmptyItem(new EmptyItem(getString(R.string.empty_list)));
        recyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        recyclerView.setAdapter(adapter);
        adapter.enableDragDrop(recyclerView);

        for (int i = 0; i < getRandom().nextInt(200) + 50; i++) {
            if (i % 2 == 0)
                adapter.add(new ExampleItem("example item " + i));
            else
                adapter.add(new TextWithButtonItem("text with button " + i));
        }
    }

    @OnClick(R.id.remove_all_items_of_a_kind)
    public void onRemoveAllItemsOfAkind() {
        adapter.removeAllItemsWithClass(ExampleItem.class);
    }

    @OnClick(R.id.remove_last_item_of_a_kind)
    public void onRemoveLastItemOfAkind() {
        adapter.removeLastItemWithClass(TextWithButtonItem.class);
    }

    @OnClick(R.id.add_item)
    public void onAddItem() {
        adapter.add(new ExampleItem("added item " + UUID.randomUUID().toString()));
    }
}

File: app/demo/src/main/java/net/gotev/recycleradapterdemo/MainActivity.kt
package net.gotev.recycleradapterdemo

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.activity_main.*
import net.gotev.recycleradapter.RecyclerAdapter
import net.gotev.recycleradapterdemo.adapteritems.LabelItem
import net.gotev.recycleradapterdemo.adapteritems.TextWithToggleItem
import net.gotev.recycleradapterdemo.adapteritems.TitleSubtitleItem
import net.gotev.recycleradapterdemo.adapteritems.leavebehind.MyLeaveBehindItem
import java.util.*


class MainActivity : AppCompatActivity() {

    private val random by lazy {
        Random(System.currentTimeMillis())
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val recyclerAdapter = RecyclerAdapter()
        recyclerAdapter.setEmptyItem(LabelItem(getString(R.string.empty_list)))

        recycler_view.apply {
            layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
            adapter = recyclerAdapter
            recyclerAdapter.enableDragDrop(this)
        }

        recyclerAdapter.add(MyLeaveBehindItem("swipe to left to leave behind", "option"))

        val items = (0..random.nextInt(200) + 50).map {
            if (it % 2 == 0)
                TitleSubtitleItem(this, "Item $it")
            else
                TextWithToggleItem("Item $it")
        }

        recyclerAdapter.add(items)

        remove_all_items_of_a_kind.setOnClickListener {
            recyclerAdapter.removeAllItemsWithClass(TitleSubtitleItem::class.java)
        }

        remove_last_item_of_a_kind.setOnClickListener {
            recyclerAdapter.removeLastItemWithClass(TextWithToggleItem::class.java)
        }

        add_item.setOnClickListener {
            recyclerAdapter.add(TitleSubtitleItem(this, "added item " + UUID.randomUUID().toString()))
        }

        search.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                recyclerAdapter.filter(search.text.toString())
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {

        R.id.sync_demo -> {
            SyncActivity.show(this)
            true
        }

        R.id.selection -> {
            SelectionActivity.show(this)
            true
        }

        R.id.selection_multi_groups -> {
            MasterSlaveGroupsActivity.show(this)
            true
        }

        else -> super.onOptionsItemSelected(item)
    }
}
